rust/qr_reader_phone/src/parser.rs

```diff
@@ -1,9 +1,67 @@
 use anyhow::anyhow;
+use nom::bits::complete::take as bit_take;
 use nom::branch::alt;
 use nom::bytes::complete::{tag, take};
+use std::convert::TryFrom;
+
+use nom::combinator::{rest, verify};
+use nom::complete::tag as bit_tag;
 use nom::error::ErrorKind;
-use nom::sequence::tuple;
-use nom::IResult;
+use nom::number::complete::be_u16;
+use nom::sequence::{preceded, tuple};
+use nom::{bits, IResult};
+
+pub(crate) struct RaptorqFrame {
+    pub(crate) size: u32,
+    pub(crate) payload: Vec<u8>,
+}
+
+impl RaptorqFrame {
+    pub(crate) fn total(&self) -> u32 {
+        // Decoding algorithm is probabilistic, see documentation of the `raptorq` crate
+        // Rephrasing from there, the probability to decode message with h
+        // additional packets is 1 - 1/256^(h+1).
+        //
+        // Thus, if there are no additional packets, probability is ~ 0.99609.
+        // If one additional packet is added, it is ~ 0.99998.
+        // It was decided to add one additional packet in the printed estimate, so that
+        // the user expectations are lower.
+        self.size / (self.payload.len() as u32) + 1
+    }
+}
+
+impl TryFrom<&Vec<u8>> for RaptorqFrame {
```

Comment:
`impl TryFrom<&[u8]>` ?

Reply:
I get
```
 the trait `From<&Vec<u8>>` is not implemented for `RaptorqFrame`
```

Reply:
Yes everywhere you call `try_from(&payload)` you would have to call `try_from(payload.as_ref())` and it should work

Reply:
Thanks! But is there a reason to prefer one over the other? Or it's more a matter of taste?

Reply:
No actually it is quite useful. When you choose `Vec<u8>` on a type-level you are binding yourself to work only with this single type. When you choose `&[u8]` you are "saying" something like "i can work with any type that can act as a contiguous piece of memory containing bytes". For idiomatic examples you can look at how the `std::io` traits are composed. For example check out the [`std::io::Write`](https://doc.rust-lang.org/std/io/trait.Write.html) trait.

Reply:
Awesome, thanks!